### ISSUE TITLE
Antlr4 jar tool generates files to an inconsistent directory

### ISSUE THREAD

**Issue by kaby76** [antlr/antlr4#3138]
I am trying to write a consistent, platform-independent and target-independent experience in testing a grammar. I just ran across an inconsistency that boggles my mind. To reproduce what I am doing, first clone and cd to the mysql code in grammars-v4 (on Windows, use Msys2 or Ubuntu, use Bash): `git clone https://github.com/antlr/grammars-v4.git; cd grammars-v4/sql/mysql`.

Then:

* On Windows, `java -jar c:/Users/kenne/Downloads/antlr-4.9.2-complete.jar Positive-Technologies/*.g4`. Observe that the generated files are in the current directory. I can use the `-o` option to place the files in Positive-Technologies/ (`java -jar c:/Users/kenne/Downloads/antlr-4.9.2-complete.jar -o Positive-Technologies Positive-Technologies/*.g4`). This is great. It does exactly what I tell it to do.
* On Ubuntu, `java -jar /tmp/antlr-4.9.2-complete.jar Positive-Technologies/*.g4`. Observe that the generated files are in _the directory Positive-Technologies_, which is different from what is done on Windows. But, worse, you cannot use the `-o .` option to change the location of the output files to the current directory. The tool ignores the `-o .` option.

A consistent presentation is first and foremost what a tool should provide, but this is not done for the Antlr tool. The Antlr tool seems to ignore the `-o` option on Ubuntu.

--Ken

**Comment by rachidlamouri** [comment]
Did you ever figure out why `-o` doesn't work as expected on Linux? I'm getting different outputs on Windows and Ubuntu for the same command https://xkcd.com/979/

**Comment by kaby76** [comment]
@rachidlamouri No, I hadn't looked at this further. I changed many of the grammars in [grammars-v4](https://github.com/antlr/grammars-v4) to follow a standardized format, which my tool and the Antlr Maven Plugin can both handle, avoiding the issue altogether. But, it seems it's still a problem: the tool should work consistently between platforms.

**Comment by rachidlamouri** [comment]
@kaby76 I found [this line](https://github.com/antlr/antlr4/blob/5d5ee8423d760470a0698ff226fd73a8a5e0889f/tool/src/org/antlr/v4/Tool.java#L841) that is causing a different behavior on Windows. It's looking for the `\` file separate instead of `/` on Windows. When it can't find the file separator it defaults to `.` which gets combined with the output directory resulting in the files going directly into the output directory.

```sh
# Windows: Places generated files directly in output/
# Ubuntu: Places generated files in output/src/
java -jar antlr-4.9.2-complete.jar src/MyLexer.g4 -o output

# Windows: Places generated files in output/src/
# Ubuntu: Source path is invalid
java -jar antlr-4.9.2-complete.jar src\\MyLexer.g4 -o output
```

For clarity: I would expect to be able to write one command that has the same behavior on both systems.

I haven't used Java much, let alone worked with platform dependent paths in Java, so it would take me a while to make a fix. Changing the behavior could also be a breaking change, so it might make sense to add another arg to input a file separator. That being said, I found the [-Xexact-output-dir](https://github.com/antlr/antlr4/blob/5d5ee8423d760470a0698ff226fd73a8a5e0889f/tool/src/org/antlr/v4/Tool.java#L141) arg which solves my use cases 🤷🏼‍♂️

**Comment by kaby76** [comment]
I'm not sure what the best fix would be, as the "-o" and the "-lib" options seem to do related things: -o = "specify output directory where all output is generated"; -lib = "specify location of grammars, tokens files". So, applying the descriptions logically to the generated/read .token files, what is the tool supposed to do for "a4 -o foo -lib bar *.g4"? Make copies in both directories? The options are unclear. This is why my driver generator stuffs everything into a flat directory and in the case of C# (Antlr4BuildTasks) uses the -o option to place the output in bin/Debug/. The options seem to have existed for quite some time ([initial version of the Antl4 tool](https://github.com/antlr/antlr4/blob/7f63527ce116f368501627f925a4db974396115c/tool/src/org/antlr/v4/Tool.java)).

**Comment by mcoblenz** [comment]
Just spent an hour debugging why Antlr was putting the generated files in the wrong directory on a Windows machine. It appears to be impossible to specify the correct output directory in a cross-platform way.
